### PR TITLE
Avoid possible UB in Clock jump callbacks

### PR DESCRIPTION
**Pull request by ivanpauno**
Probably fixes https://github.com/ros2/rclcpp/issues/938.

Anyways, "possible UB" should never be part of a TODO comment:
https://github.com/ros2/rclcpp/blob/efbce4a11be748baf9d7dd99706b1004f76671bc/rclcpp/src/rclcpp/clock.cpp#L131

### CODE REVIEW

File: rclcpp/src/rclcpp/clock.cpp [hunk 1]
```diff
@@ -30,29 +32,36 @@ JumpHandler::JumpHandler(
...
+  delete rcl_clock;
 }
 
-Clock::~Clock()
+Clock::Clock(rcl_clock_type_t clock_type)
 {
-  auto ret = rcl_clock_fini(&rcl_clock_);
+  allocator_ = rcl_get_default_allocator();
+  auto clock_unique_ptr = std::make_unique<rcl_clock_t>();
+  rcl_clock_ = std::shared_ptr<rcl_clock_t>(clock_unique_ptr.release(), rcl_clock_deleter);
```
Comment (hidmic) [inline rclcpp/src/rclcpp/clock.cpp]: @ivanpauno fyi: I see what you're doing here, but it's not better than `new rcl_clock_t()` right there on line 50..

Comment (wjwwood) [reply rclcpp/src/rclcpp/clock.cpp]: Yeah, the temporary `unique_ptr` is only useful if some exception may occur between allocating and initializing the clock. Unless I miss something, I agree with hidmic that `new rcl_clock_t` will work just as well.

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8936/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4707/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7308/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8843/)

File: rclcpp/src/rclcpp/clock.cpp [hunk 2]
```diff
@@ -30,29 +32,36 @@ JumpHandler::JumpHandler(
...
   }
+  delete rcl_clock;
 }
 
-Clock::~Clock()
+Clock::Clock(rcl_clock_type_t clock_type)
 {
-  auto ret = rcl_clock_fini(&rcl_clock_);
+  allocator_ = rcl_get_default_allocator();
+  auto clock_unique_ptr = std::make_unique<rcl_clock_t>();
```
Comment (clalancette) [inline rclcpp/src/rclcpp/clock.cpp]: Did you mean to remove this line as well?  I don't see it referenced anywhere.

Comment (ivanpauno) [reply rclcpp/src/rclcpp/clock.cpp]: Ups, forgot to delete that. Thanks!

**Comment by ivanpauno** [comment]
Extra run with the deleted line:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8937/)

**Comment by ivanpauno** [comment]
The windows failure is flaky and unrelated:
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8845/)

**Pull request merged**